cmake/precompiled_header.cmake

```diff
@@ -2,8 +2,7 @@
 # header name as input. The function will generate a .cpp file that includes the header and is used
 # to generate the precompiled header; this source file is added to the target's sources.
 function(target_precompiled_header target_name header_name)
-    if (MSVC)
-
+    if (MSVC AND CMAKE_VS_PLATFORM_TOOLSET)
```

Comment:
PCHs should work under ninja, but the way it's built is currently broken (it writes a file named ${header_base_name}.cpp. Problem is that there are multiple pchs called pch.h in WinML, which when built in parallel leads to Ninja giving permission denied or internal compiler errors). So this was already broken, but happened to work under msbuild. We should fix the repo to remove headers with the same name.

Reply:
curious - why can't headers with the same name exist in different folders? that seems like it should be supported.

Reply:
It could be supported, it's just that right now it writes pch.cpp to `${CMAKE_CURRENT_BINARY_DIR}/${header_base_name}.cpp` ([at this line](https://github.com/microsoft/onnxruntime/pull/2790/files/42f701c47df04d1f679adbe086214e59520b4009#diff-17a2406566fef690dd4b786796e16ef3R12)) and compiles it, so two headers with the same base name go to the same cpp file. If we want to support it we'd have to change precompiled_header.cmake to write to unique file names or different directories.

Reply:
Answering the question whether the build output is the same: I'm comparing the linker command line for VS/msbuild and ninja and they aren't exactly the same. Looking at issues and CMake source code, it seems like some are added by the msbuild CMake generator, and some are added to the VS project by default.

The full command line for linking onnxruntime is:

Ninja:
`C:\PROGRA~2\MICROS~2\2019\ENTERP~1\VC\Tools\MSVC\1424~1.283\bin\Hostx64\x64\link.exe /nologo CMakeFiles\onnxruntime.dir\generated_source.c.obj /out:onnxruntime.dll /implib:onnxruntime.lib /pdb:onnxruntime.pdb /dll /version:1.1 /machine:x64 /WX /debug /INCREMENTAL -DEF:D:/onnxruntime/build/Debug/onnxruntime_dll.def /DELAYLOAD:DirectML.dll /DELAYLOAD:d3d12.dll /DELAYLOAD:dxgi.dll onnxruntime_session.lib onnxruntime_providers_dml.lib windowsapp.lib winml_adapter.lib onnxruntime_optimizer.lib onnxruntime_providers.lib onnxruntime_util.lib onnxruntime_framework.lib onnxruntime_graph.lib onnxruntime_common.lib onnxruntime_mlas.lib onnx\onnx.lib onnx\onnx_proto.lib external\protobuf\cmake\libprotobuf-lited.lib external\re2\re2.lib advapi32.lib Shlwapi.lib Dbghelp.lib d3d12.lib dxgi.lib packages\DirectML.0.0.1\bin\x64\DirectML.lib kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib /MANIFEST /MANIFESTFILE:CMakeFiles\onnxruntime.dir/intermediate.manifest CMakeFiles\onnxruntime.dir/manifest.res`

VS:
`C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.24.28314\bin\HostX64\x64\link.exe /ERRORREPORT:QUEUE /OUT:"D:\onnxruntime\build\msbuild\Debug\onnxruntime.dll" /INCREMENTAL /NOLOGO /WX Debug\onnxruntime_session.lib Debug\onnxruntime_providers_dml.lib windowsapp.lib Debug\winml_adapter.lib Debug\onnxruntime_optimizer.lib Debug\onnxruntime_providers.lib Debug\onnxruntime_util.lib Debug\onnxruntime_framework.lib Debug\onnxruntime_graph.lib Debug\onnxruntime_common.lib Debug\onnxruntime_mlas.lib onnx\Debug\onnx.lib onnx\Debug\onnx_proto.lib "external\protobuf\cmake\Debug\libprotobuf-lited.lib" external\re2\Debug\re2.lib advapi32.lib Shlwapi.lib Dbghelp.lib d3d12.lib dxgi.lib packages\DirectML.0.0.1\bin\x64\DirectML.lib kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib Delayimp.lib /DEF:"D:/onnxruntime/build/msbuild/onnxruntime_dll.def" /DELAYLOAD:DirectML.dll /DELAYLOAD:d3d12.dll /DELAYLOAD:dxgi.dll /MANIFEST /MANIFESTUAC:"level='asInvoker' uiAccess='false'" /manifest:embed /DEBUG /PDB:"D:/onnxruntime/build/msbuild/Debug/onnxruntime.pdb" /SUBSYSTEM:CONSOLE /TLBID:1 /DYNAMICBASE /NXCOMPAT /IMPLIB:"D:/onnxruntime/build/msbuild/Debug/onnxruntime.lib" /MACHINE:X64  /machine:x64 /DLL onnxruntime.dir\Debug\generated_source.obj`

The differences are:

Ninja has:
`/version:1.1 /MANIFESTFILE:CMakeFiles\onnxruntime.dir/intermediate.manifest CMakeFiles\onnxruntime.dir/manifest.res`

VS has:
`/ERRORREPORT:QUEUE Delayimp.lib /MANIFESTUAC:"level='asInvoker' uiAccess='false'" /manifest:embed /SUBSYSTEM:CONSOLE /TLBID:1 /DYNAMICBASE /NXCOMPAT`

/ERRORREPORT:QUEUE doesn't interfere with the build, and /SUBSYSTEM:CONSOLE /TLBID:1 /DYNAMICBASE /NXCOMPAT are defaults so shouldn't interfere. CMakeFiles\onnxruntime.dir/intermediate.manifest in the ninja build is a XML with `<requestedExecutionLevel level='asInvoker' uiAccess='false' />`, which matches the options passed in the command line by msbuild (`/MANIFESTUAC:"level='asInvoker' uiAccess='false'"`), so should achieve the same. So the relevant options are:

* Ninja added /version:1.1
* msbuild has Delayimp.lib and /manifest:embed

I'm debugging a bit more to find out where Delayimp.lib came from; it's not added in any of our CMake files, so it's either added by the VS generator of CMake, or added by VS by default.